test/test_overrides.py

```diff
@@ -461,24 +488,99 @@ def test_user_implementation_raises(self):
         with self.assertRaises(ValueError):
             quux(t1)
 
+    def test_tensor_subclass_propagation(self):
+        t1 = torch.tensor([5])
+        t2 = torch.tensor([6])
+
+        s1 = SubTensor2([5])
+        s2 = SubTensor2([6])
+
+        ss1 = SubSubTensor2([5])
+        ss2 = SubSubTensor2([6])
+
+        sn1 = SubTensor3([5])
+        sn2 = SubTensor3([6])
+
+        # Check that leaf subclass is kept regardless of order
+        self.assertTrue(isinstance(s1 + t2, SubTensor2))
+        self.assertTrue(isinstance(t1 + s2, SubTensor2))
+        self.assertTrue(isinstance(s1 + s2, SubTensor2))
+
+        # Check indexing subclass is kept
+        self.assertTrue(isinstance(s1[0], SubTensor2))
+
+        # Check case for subclass of subclass.
+        self.assertTrue(isinstance(ss1 + ss2, SubSubTensor2))
+        self.assertTrue(isinstance(ss1 + s2, SubSubTensor2))
+        self.assertTrue(isinstance(s1 + ss2, SubSubTensor2))
+        self.assertTrue(isinstance(ss1 + ss2, SubSubTensor2))
+        self.assertTrue(isinstance(ss1 + t2, SubSubTensor2))
+        self.assertTrue(isinstance(t1 + ss2, SubSubTensor2))
+        self.assertTrue(isinstance(ss1[0], SubSubTensor2))
+
+        # Make sure unrelated class trees are not merged.
+        with self.assertRaises(TypeError):
+            s1 + sn2
+        with self.assertRaises(TypeError):
+            sn1 + s2
+
+
 def generate_tensor_like_override_tests(cls):
     def test_generator(func, override):
+        # If func corresponds to a torch.Tensor method or property.
+        if HANDLED_FUNCTIONS_NAMESPACES[func] is torch.Tensor or func.__name__ == "__get__":
+            # Generate an instance by using SubTensor,
+            instance_gen = lambda: SubTensor([5])
+        else:
+            # Otherwise, TensorLike.
+            instance_gen = lambda: TensorLike()
         args = inspect.getfullargspec(override)
         nargs = len(args.args)
         if args.defaults is not None:
             nargs -= len(args.defaults)
-        func_args = [TensorLike() for _ in range(nargs)]
+        func_args = [instance_gen() for _ in range(nargs)]
         if args.varargs is not None:
-            func_args += [TensorLike(), TensorLike()]
+            func_args += [instance_gen(), instance_gen()]
 
         def test(self):
-            self.assertEqual(func(*func_args), -1)
-
+            ret = func(*func_args)
+            # ret is None for certain protocols, e.g., `__weakref__` and `__setitem__`
+            # This is currently the best check but doesn't work for, for example,
+            # Tensor.__add__ because it redirects to Tensor.add.
+            if ret is None:
+                self.assertTrue(HANDLED_FUNCTIONS_WRAPPERS[func]._triggered)
```

Comment:
Regarding https://github.com/pytorch/pytorch/pull/37091#discussion_r431388402

This will be the only test needed if we go the "maximally flexible" route.